src/repo.rs

```diff
@@ -732,7 +766,7 @@ impl DocumentInfo {
         }
         let waker = Arc::new(RepoWaker::Storage(wake_sender.clone(), document_id));
         self.state.poll_pending_save(waker);
-        self.patches_since_last_save = 0;
+        self.last_heads = new_heads;
```

Comment:
Can this lead to data loss? We update the heads of the last time we saved before the storage future has resolved, which means the storage future can fail for some reason (or more likely, be stuck in some kind of deadlock) and then then next time we save we only save since the heads which _havent_ made it to disk.

Reply:
I'm not sure about whether it can lead to data loss, but in any case it does make sense to store the heads together with the future onto the `DocState`, and then `note_changes` could use the heads of the last (fut, heads) pair? I think it's a good follow-up, because it's a complicated change to make.